### PR TITLE
Reduce duplications in ubuntu-based Dockerfiles 

### PR DESCRIPTION
**Pull request by KaiSzuttor**

### CODE REVIEW

File: docker/Dockerfile-ubuntu-20.04 [hunk 1]
```diff
@@ -1,44 +1,11 @@
...
-    python3-pip \
-    python3-requests \
-    python3-setuptools \
-    libhdf5-openmpi-dev \
-    libhdf5-openmpi-103:amd64 \
-    libhdf5-103:amd64 \
-    libblas-dev \
-    liblapack-dev \
-    vim \
-    gdb \
```
Comment (jngrad) [inline docker/Dockerfile-ubuntu-20.04:35]: why remove GDB?

File: docker/Dockerfile-ubuntu-20.04 [hunk 2]
```diff
@@ -1,44 +1,11 @@
...
-    ccache \
-    autoconf \
-    gfortran \
-    build-essential \
-    pkg-config \
-    openmpi-bin \
-    libopenmpi-dev \
-    clang-9 \
-    clang-tidy-9 \
-    clang-format-9 \
```
Comment (jngrad) [inline docker/Dockerfile-ubuntu-20.04:16]: why remove clang? this image is no longer going to replace `clang:6.0`?

File: docker/Dockerfile-ubuntu-cuda-10.1 [hunk 3]
```diff
@@ -1,40 +1,24 @@
...
-    build-essential \
-    cmake ccache \
-    gfortran \
-    pkg-config \
-    autoconf \
-    openmpi-bin libopenmpi-dev \
-    libfftw3-dev \
-    libboost-dev libboost-serialization-dev libboost-mpi-dev libboost-filesystem-dev libboost-test-dev \
-    libgsl-dev \
-    cython3 python3 python3-dev python3-pip python3-setuptools python3-six \
```
Comment (jngrad) [inline docker/Dockerfile-ubuntu-cuda-10.1:14]: we don't need `six`?

Comment (KaiSzuttor) [reply docker/Dockerfile-ubuntu-cuda-10.1:14]: I could not find any direct usage

File: docker/Dockerfile-ubuntu-20.04 (hunk 2 above)
Comment (KaiSzuttor) [reply docker/Dockerfile-ubuntu-20.04:16]: will fix

File: docker/Dockerfile-ubuntu-20.04 (hunk 1 above)
Comment (KaiSzuttor) [reply docker/Dockerfile-ubuntu-20.04:35]: will fix

File: docker/ubuntu-packages.txt [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+build-essential
+ccache
+cmake
+curl
+cython3
+doxygen
+openmpi-bin libopenmpi-dev
+gfortran
+git
+graphviz
```
[the remark this answers is not in the document]
Comment (KaiSzuttor) [reply docker/ubuntu-packages.txt:12]: will fix

File: docker/ubuntu-packages.txt [hunk 5]
```diff
@@ -0,0 +1,29 @@
+apt-utils
+autoconf
+build-essential
+ccache
+cmake
+curl
+cython3
+doxygen
```
[the remark this answers is not in the document]
Comment (KaiSzuttor) [reply docker/ubuntu-packages.txt:8]: will fix

File: docker/Dockerfile-ubuntu-18.04 [hunk 6]
```diff
@@ -1,33 +1,11 @@
...
-    ccache \
-    curl \
-    jq \
-&& apt-get clean \
-&& rm -rf /var/lib/apt/lists/*
+COPY ubuntu-packages.txt /tmp
+RUN apt-get update && xargs -a /tmp/ubuntu-packages.txt apt-get install -y
+RUN apt-get install -y libhdf5-openmpi-dev libhdf5-openmpi-100:amd64 libhdf5-100:amd64
+RUN apt-get clean
+RUN rm -rf /var/lib/apt/lists/*
```
[the remark this answers is not in the document]
Comment (KaiSzuttor) [reply docker/Dockerfile-ubuntu-18.04:8]: will fix

**Comment by RudolfWeeber** [comment]
My takeaway from yesterday's discussion was that I would work on Ubuntu 20.4 images before other docker related changes happen. I did not start, so no damage is done, but we need to coordinate this.

My suggestion is that
* this PR gets finished and merged
* then, I'll do the Ubuntu 20 work
* once that's done, toher docker changes can happen

**Comment by KaiSzuttor** [comment]
sorry, you are right. It would probably also be a good idea to use issues and assignments in this repository.

**Comment by jngrad** [comment]
I also thought I would be the one to do this simplification of the packages. Haven't worked on it yet, so no damage from my side either.

**Comment by KaiSzuttor** [comment]
however, merging Dockerfiles is trivial since they are in the order of 1e1 LOC where 90 percent are packagenames

**Comment by RudolfWeeber** [comment]
With regares to package selection:
I'd suggest
* kepping the package lists for different containers as similar as possible. Only when a container is supposed to test successful build in the absence of a package, do we need to actually exclude that.
* Gdb, vim, perf and other helpful dev tools should be present in all containers, IMO

File: docker/Dockerfile-ubuntu-19.04 [hunk 7]
```diff
@@ -1,33 +1,11 @@
...
-    liblapack-dev \
-    autoconf \
-    vim \
-    ccache \
-    curl \
-    jq \
-&& apt-get clean \
-&& rm -rf /var/lib/apt/lists/*
+COPY ubuntu-packages.txt /tmp
+RUN apt-get update && xargs -a /tmp/ubuntu-packages.txt apt-get install -y && \
```
Comment (jngrad) [inline docker/Dockerfile-ubuntu-19.04:5]: The older syntax followed the docker best practices by putting the `&&` at the beginning of lines to make it easier to distinguish which line continuations were shell commands (it starts with `&&`) and which lines continuations were list of packages (it doesn't start with `&&`). Why move the `&&` to the end of lines?

File: docker/Dockerfile-ubuntu-20.04 [hunk 8]
```diff
@@ -1,44 +1,12 @@
...
-    libfftw3-dev \
-    libboost-dev libboost-serialization-dev libboost-mpi-dev libboost-filesystem-dev libboost-test-dev \
-    libgsl-dev \
-    cython3 \
-    python3 \
-    python3-numpy \
-    python3-scipy \
-    python3-h5py \
-    python3-vtk7 \
-    python3-pip \
```
Comment (jngrad) [inline docker/Dockerfile-ubuntu-20.04:26]: but we need pip in this image

File: docker/Dockerfile-ubuntu-cuda-9.0 [hunk 9]
```diff
@@ -1,33 +1,10 @@
...
-    apt-utils \
-    build-essential \
-    curl \
-    lcov \
-    pkg-config \
-    openmpi-bin libopenmpi-dev \
-    libfftw3-dev \
-    libboost-dev libboost-serialization-dev libboost-mpi-dev libboost-filesystem-dev libboost-test-dev \
-    libgsl-dev \
-    python3 python3-numpy python3-scipy python3-h5py python3-pip \
```
Comment (jngrad) [inline docker/Dockerfile-ubuntu-cuda-9.0:13]: please remove the pip command and the ENV PATH command at the bottom of this file

File: docker/Dockerfile-ubuntu-20.04 (hunk 8 above)
Comment (jngrad) [reply docker/Dockerfile-ubuntu-20.04:26]:
```
/bin/sh: 1: pip3: not found
Pull request: not pushing to registry.
```
Yet the checkmark is green...

Comment (KaiSzuttor) [reply docker/Dockerfile-ubuntu-20.04:26]: where is it green?
https://github.com/KaiSzuttor/docker/runs/557442779?check_suite_focus=true

Comment (KaiSzuttor) [reply docker/Dockerfile-ubuntu-20.04:26]: ah that's weird

Comment (jngrad) [reply docker/Dockerfile-ubuntu-20.04:26]: https://github.com/espressomd/docker/runs/557443033

**Comment by KaiSzuttor** [comment]
i checked the images locally. done

**Pull request merged**